### ISSUE TITLE
Fix accounts-twitter and accounts-facebook to not force Blaze into an app

### ISSUE THREAD

**Issue by queso** [meteor/meteor#7715]
This is an open issue to discuss the idea of getting Blaze deps out of packages like accounts-twitter and accounts-facebook so that we don't force Blaze on people who may be using React/Angular/Vue, etc.

This is an extension of a discussion started on the forum and intended to discuss ideas and whatnot around this.

**Comment by queso** [comment]
So @SachaG suggested that we could just make accounts-ui and Blaze a weak dependency for accounts-twitter, et al. Not sure that will work because accounts-facebook just relies on accounts-oauth and accounts-base (https://github.com/meteor/meteor/blob/devel/packages/accounts-facebook/package.js). Going to dig deeper.

**Comment by queso** [comment]
So it does look like accounts-base calls in Blaze as a weak dependency (https://github.com/meteor/meteor/blob/devel/packages/accounts-base/package.js#L30), so I have no idea why it is coming along all the way atm. My test project doesn't have anything else needing Blaze. My guess is that some other package is bringing it in that accounts-base relies on.

**Comment by SachaG** [comment]
I could be wrong, but I think it's: https://github.com/meteor/meteor/blob/devel/packages/twitter/package.js#L8

**Comment by queso** [comment]
Nice, I just found that. It looks like the accounts-* oauth packages all rely on a package like the twitter one above. This is forcing Blaze in even thought it is a weak dep for accounts-base.

**Comment by queso** [comment]
So it looks like the underlying `twitter` package and it's ilk handle two things and I think those things need to be broken out into two packages...

1) It handles the actual oAuth calls for each service
2) It contains the html/js to power the 'configuration' dialog if you are using accounts-ui.

@stubailo: Do you have a preference on how this should be handled to extract these packages apart? My gut says the functionality of #2 should be moved off to new packages... And the current `twitter`, `facebook`, and those types of packages should continue to be required by accounts-twitter.

The second option is to move the code related to #1 into accounts-twitter directly and make `twitter` a weak dep that people can install if they want the configuration.

**Comment by stubailo** [comment]
Hmm, do we think anyone would ever use the Facebook package without accounts-Facebook?

**Comment by Primigenus** [comment]
You know what would be great? If I add `accounts-facebook` and `accounts-ui` and it just knows which view layer my app is already using and composes the right one for me.

**Comment by queso** [comment]
> You know what would be great? If I add accounts-facebook and accounts-ui and it just knows which view layer my app is already using and composes the right one for me.

I would rather keep this focused on achievable, smaller goals.

**Comment by queso** [comment]
> Hmm, do we think anyone would ever use the Facebook package without accounts-Facebook?

I guess it is possible. Looking at the two packages, it is an odd way to divide the code. The code for dealing with oAuth lives in `facebook`, but the code for calling the `Meteor.loginWithFacebook` lives in `accounts-facebook`. Those two things feel like they should live together perhaps, in my mind.

The existance of the configuration html inside the `facebook` package is the real problem here. There are plenty of ways to configure the oAuth services without needing the UI.

To answer your original question: maybe? If they do install facebook without accounts-facebook, it would be to get oAuth support without needing `Meteor.loginWithFacebook`...

**Comment by MeKarina** [comment]
i don't think making blaze related package a weak dependency would solve this issue because all meteor project still include blaze on server due to `boilerplate` package.... before we deal with that issue,, weak dependency is not a solution i guess... see my comment on this issue... https://github.com/meteor/meteor/issues/6844#issuecomment-242799957

`account-*` use blaze for ui and give ability to fill configuration (appId/appSecret) on web page...

the most easiest way to remove blaze dependency is to create new packages... something like... `account-*-without-ui`... these package basically `account-*` package without configuration and ui... this way,,, `account-*-without-ui` will agnostic to and won't depend to any ui...

then,,
install `account-*` if developer want account with blaze ui and ability to fill configuration on web page (appId/appSecret)..
or,,
install `account-*-without-ui` if developer want account with no ui dependency and manual config (appId/appSecret)..

```js
ServiceConfiguration.configurations.update(
  {"service":"google"},
  {$set: {
    "clientId": "yyy",
    "secret": "zzz"}
  },
  {upsert:true}
);
```

i tested `account-*-without-ui` on local packages folder and working properly...

if someone want to proceed this into pr,,, this actually pretty easy,,, no need to add/change code,, just need to change the structure of what depend to what (`package.js`) and of course this requere to break `* package` into piece of packages and move some files to new package for the sake of code sharing and maintenance sharing and easier to move to npm in the future...

**the other solution...**

rewrite the whole account system into es6 class module system... with this,,, u could import what u need... if u only need to authenticate,,then u just import `loginWith*` without importing configuration files... then the blaze won't be imported i guess...

karina..

**Comment by queso** [comment]
> i don't think making blaze related package a weak dependency would solve this issue because all meteor project still include blaze on server due to boilerplate package.... before we deal with that issue,, weak dependency is not a solution i guess... see my comment on this issue... #6844 (comment)

I disagree, having a weak dependency would solve it, I could care less about the server, I am talking about having Blaze shoved into my client bundle, it affects page load times due to files sizes and bandwidth use.

> account-* use blaze for ui and give ability to fill configuration (appId/appSecret) on web page...

> the most easiest way to remove blaze dependency is to create new packages... something like... account-*-without-ui... these package basically account-* package without configuration and ui... this way,,, account-*-without-ui will agnostic to and won't depend to any ui...

> then,,
install account-* if developer want account with blaze ui and ability to fill configuration on web page (appId/appSecret)..
or,,
install account-*-without-ui if developer want account with no ui dependency and manual config (appId/appSecret)..

See, that actually wouldn't work. We are dealing with two layers of issues here. There is `accounts-facebook` and `facebook` packages that are needed to power oAuth. It is the `facebook` package that has the UI pieces in it purely for configuration ui with accounts-ui. I feel like the oAuth specific code should be moved out of `facebook` and into `accounts-facebook`, for example.

**Comment by queso** [comment]
For completeness sake, I am talking about this code: https://github.com/meteor/meteor/blob/devel/packages/facebook/facebook_client.js

and this code:
https://github.com/meteor/meteor/blob/devel/packages/facebook/facebook_server.js

When talking about oAuth stuff that lives in the `facebook` package instead of `accounts-facebook`.

**Comment by queso** [comment]
> i tested account-*-without-ui on local packages folder and working properly...

@MeKarina did you check your list of js files that are getting push or your .meteor/versions file, I bet you are getting blaze still...

**Comment by stubailo** [comment]
I think it would be fine to just move everything into the `accounts-facebook` package, and then have `facebook` be just the UI part. For backcompat, let's make sure `facebook` still re-exports all of the stuff it used to, and maybe warns you if you don't have `accounts-facebook` installed?

I think that will cover all of the bases for someone who was using `facebook` independently, at least they will be notified they need to change their code.

**Comment by MeKarina** [comment]
@queso yepp as i stated on https://github.com/meteor/meteor/issues/6844#issuecomment-242799957 they're still staying on versions file,,, but i don't find on client bundle... 

i u want to remove blaze completely,,, include on server,, u should think how to remove blaze from..
1. https://github.com/meteor/meteor/blob/devel/packages/boilerplate-generator/boilerplate_web.browser.html
2. https://github.com/meteor/meteor/blob/devel/packages/boilerplate-generator/boilerplate_web.cordova.html
3. https://github.com/meteor/meteor/blob/devel/packages/boilerplate-generator/boilerplate-generator.js#L34-L35

the blaze on versions file used by these 3 files which are server only and if me,,, i couldn't care less...

**Comment by stubailo** [comment]
@MeKarina want to remove the `boilerplate-generator` dependency? I feel like you could replace it with handlebars from NPM or something - it's literally just rendering a template to a string.

**Comment by laosb** [comment]
I think having only ui things in facebook can be confusing. We should directly deprecate that package (but exports what it used to export) and print a message when someone use it. Configuration ui should be in a new package called blaze-accounts-config-*(or something like that).

**Comment by stubailo** [comment]
That works for me as well - probably isn't much more work to do it that way, and is definitely more correct.

**Comment by MeKarina** [comment]
@stubailo if it's replaced with handlebars,,, `static-html` package won't working anymore..??

**Comment by stubailo** [comment]
@MeKarina why not?

**Comment by MeKarina** [comment]
dunno.. just asking.. =))

**Comment by stubailo** [comment]
You can easily make static-html work without Blaze, but I think that should be a separate discussion.

**Comment by queso** [comment]
I agree with @laosb here, likely better to make a new package with a solid name and deprecate the old one.

**Comment by MeKarina** [comment]
yepp agree that should be easier... 

i'm not blaze user so i shouldn't care... but the change made could be confusing for blaze users... not to mention that i just visit forums and read a lot complain from blaze users...

this psychological aspect might worth to consider as well...

also blaze based account packages like useraccount...

**Comment by laosb** [comment]
So that sounds like:

1. Move everything except configuration UI from package `{provider}` to `accounts-{provider}`
2. Move configuration UI to `blaze-accounts-config-ui-*`, and `accounts-*` can have a weak dependency on it.
3. Deprecate `{provider}`.

**Comment by queso** [comment]
@laosb: that sounds spot on from what I am getting from the conversation. Sound good to you too @stubailo?

**Comment by stubailo** [comment]
Sounds perfect, let's do it!

**Comment by laosb** [comment]
Working on it!

**Comment by laosb** [comment]
@stubailo Updated the solution above to be more detailed. How about it?

**Comment by stubailo** [comment]
Still sounds good!

**Comment by laosb** [comment]
Do we need to keep a copy of configuration UI templates inside `facebook` to keep the compatibility?

**Comment by stubailo** [comment]
I think `facebook` can just have a dependency on the UI package?

**Comment by laosb** [comment]
Oh, yeah.

**Comment by laosb** [comment]
@stubailo I submitted #7728 for this.

**Comment by tmeasday** [comment]
Apologies for coming late to the party here folks but although I agree with the general thrust of this, I think we should do it slightly differently.

1. I think we should keep the `facebook` package. I'm not sure if people use it much (for doing facebook oauth without using the accounts system), I don't see a compelling reason to break backcompat here.

2. I'd prefer a solution where if you are using `accounts-ui` and you add `accounts-facebook` you get the UI configuration without an extra step (which would break a *whole* lot of existing tutorials etc).

  What do we think about just adding the facebook config template (and eventually the other core `accounts-X` UI templates) to `accounts-ui`? I know it's bloat but that package is rarely used in production and templates are pretty tiny.

  It also means treating the "core" oauth packages differently, but I'm OK with that.

**Comment by tmeasday** [comment]
Removing `pull-requests-encouraged` until we sort out the exact implementation.

**Comment by tmeasday** [comment]
@laosb said (on the PR):

> I guess that won't make someone who only use accounts-password happy. It's small but still too much for someone don't use it. Actually I'm using accounts-ui in production with our own styles.

> I would suggest printing a message when accounts-{provider} find accounts-ui to tell users add the small ui package. In fact even they do use accounts-ui and accounts-{provider}, many of them will still configure it through environment variables or something like that, for it's easier in production.

**Comment by tmeasday** [comment]
It would probably be possible to add something like:

```js
// in package.js
api.use(['accounts-ui', 'accounts-facebook-ui', {weak: true});

// in a js file
if (Package['accounts-ui'] && !Package['accounts-facebook-ui']) {
  // log a warning
}
```

**Comment by mitar** [comment]
> I know it's bloat but that package is rarely used in production and templates are pretty tiny.

So you would remove separation between packages and now `accounts-ui` would contain all configs for other packages? Isn't this ugly? Or am I misunderstanding something.

So I agree that settings template for Twitter and Facebook are used in production rarely as it is, but I am not sure if `accounts-ui` is used rarely.

**Comment by laosb** [comment]
So we don't deprecate {provider} and just move configuration ui templates to a separated package? Actually I didn't see the sense of a package only do the half of the whole thing and actually we do make facebook exports what it used to export. It is back compatible.

**Comment by tmeasday** [comment]
@laosb the idea is that people that aren't using the accounts system may still want to do oauth with FB, and the `facebook` package helps with that. I would say it's a rare use case, and if we were starting from scratch, I wouldn't bother with it, but it seems like there's no reason to deny this use case at this point.

**Comment by tmeasday** [comment]
> So I agree that settings template for Twitter and Facebook are used in production rarely as it is, but I am not sure if accounts-ui is used rarely.

OK, I think I'm convinced.

So my proposed design now would be 
a) Create a new `accounts-ui-facebook-config` (I think this is a more appropriate name, given `accounts-ui` is blaze always) that takes the config stuff from `facebook`

b) Log a warning if you are using `accounts-ui` and `accounts-facebook` but not `accounts-ui-facebook-config` (ala https://github.com/meteor/meteor/issues/7715#issuecomment-243606006)

**Comment by mitar** [comment]
Why we do not create two packages `accounts-ui-facebook-config` and `facebook-core`, and then make `facebook` depend on `accounts-ui-facebook-config` and `facebook-core`?

**Comment by tmeasday** [comment]
@mitar I don't think there's a use-case that makes sense for using `facebook` + `accounts-ui` but not `accounts-facebook`. Am I missing something?

**Comment by mitar** [comment]
I do not think so, it was just to make it backwards compatible. But you are right, warning is probably also good.

**Comment by mitar** [comment]
(Or exception?)

**Comment by tmeasday** [comment]
I think there is a legitimate use case for wanting the accounts UI and facebook oauth, but not the facebook configuration screen, so I'd say a warning

**Comment by MeKarina** [comment]
i add more explanation to my original proposal... =))

**current structure...**
```sh
accounts-{service}
=> accounts-base
=> underscore
=> random
=> accounts-oauth
=> {service}
- {service}.js

{service}
=> oauth2
=> oauth
=> http
=> underscore
=> service-configuration
=> random
=> templating
- {service}_server.js
- {service}_client.js
- {service}_configure.html
- {service}_configure.js
```

**proposed structure....**
```sh
accounts-{service}
=> accounts-base
=> underscore
=> random
=> accounts-oauth
=> {service}
- {service}.js

accounts-{service}-no-ui
=> accounts-base
=> underscore
=> random
=> accounts-oauth
=> {service}-oauth
- {service}.js

{service}
=> {service}-oauth
=> {service}-config

{service}-oauth
=> oauth2
=> oauth
=> http
=> underscore
=> random
- {service}_server.js
- {service}_client.js

{service}-config
=> templating
=> service-configuration
- {service}_configure.html
- {service}_configure.js
```

**usage...**
- blaze..
`meteor add accounts-{service}`

- non blaze...
`meteor add accounts-{service}-no-ui`

**bonus package... =))**
`meteor add {service}-oauth`
if someone need {service} oauth with manual configuration without ui for adding clientId and of course without adding blaze package...

**pros** 
- no breaking change for `meteor add accounts-{service}`.. i guess...
- no breaking change for `meteor add {service}`.. maybe...
- safer from unexpected behavior or unexpected usage.. maybe..

**cons**
- more complex... or maybe no because i just simply break {service} package into {service}-oauth and {service}-config...

**Comment by laosb** [comment]
Maybe `facebook-core` is a better name for a package than just `facebook`. That's why I suggest deprecating `facebook`.

**Comment by tmeasday** [comment]
I think @MeKarina might be onto something, although I think it's a little level too complicated - I don't think we need the flexibility to install `{service}-config` without `accounts-{service}`

So what about this proposal:

- `accounts-{service}` - `accounts-{service}-no-ui` + the configuration UI
- `accounts-{service}-no-ui` - `{service}` + accounts integration
- `{service}` - just the oauth bit [we could rename this to `{service}-oauth` potentially.

What I like about this:
 - We keep back-compat completely (except that `{service}` loses the useless config UI), so all existing documentation / demos work

What I don't like:
 - `accounts-{service}-no-ui` is kind of wordy
 - People will end up with the UI if they don't think too hard, which might be a poor default.

**Comment by MeKarina** [comment]
@tmeasday 

> loses the useless config UI

i don't think it is useless... it create wow impression for potential new user... including me when i tried meteor todo demo for the first time... =))

> `accounts-{service}-no-ui` is kind of wordy

agree,, =P

**Comment by MeKarina** [comment]
this a bit better maybe...

```js
accounts-{service}
=> accounts-{service}-no-ui
=> {service}-config

accounts-{service}-no-ui
=> accounts-base
=> underscore
=> random
=> accounts-oauth
=> {service}-oauth
- {service}.js

{service}
=> {service}-oauth
=> {service}-config

{service}-oauth
=> oauth2
=> oauth
=> http
=> underscore
=> random
=> service-configuration
- {service}_server.js
- {service}_client.js

{service}-config
=> templating
- {service}_configure.html
- {service}_configure.js
```

usages are the same...

**Comment by tmeasday** [comment]
@MeKarina 

> i don't think it is useless... it create wow impression for potential new user... including me when i tried meteor todo demo for the first time... =))

Oh, if it was unclear, I meant that it's useless if you aren't using accounts (i.e `facebook` but not `accounts-facebook`).

**Comment by mitar** [comment]
BTW, looking into this I realized that anyone can configure services even in production if they were left unconfigured. It would be great if we could remove that feature as well: https://github.com/meteor/meteor/issues/7745 So that if you want this simple configuration + configuration template you should add some package. But otherwise `service-configuration` package should just give you a server side collection.

**Comment by MeKarina** [comment]
Could it be possible for developer to left their login unconfigured in production..?? Signup/login is the very first gate no..?? It might be the very first they were tested before going to production..??

**Comment by mitar** [comment]
I think when somebody is just deploying an app this could happen. Just imagine app like Rockat.Chat. (An abstract case, I am not sure if they really do it like that, but it is an example of an app in Meteor.) You deploy a Docker container. And you do not configure Facebook integration because you do not want it. And they do not use accounts-ui for UI, so there is no configuration prompt displayed. But their method is exposed so somebody can still configure it by calling the method directly the first time.

**Comment by laosb** [comment]
Basically yes, many people are doing like this, which is a (maybe) potential security issue. (Please keep further discussions in #7745, since it's not the same issue.)

**Comment by laosb** [comment]
IMO the proposal by @MeKarina is still too complex. Also, it shouldn't be `{service}-config` which is also misleading, since they're templates for `accounts-facebook`.

Now seems whether the configuration templates should be default in `accounts-{service}`. I personally like a view-layer-free `accounts-{service}`, since it's easier to understand and clear even if you just look at the package names to guess what to do. The name, `accounts-{service}`, doesn't suggest some templates in it. We'll just need to update `meteor add accounts-facebook` to meteor add accounts-facebook accounts-facebook-config-ui` to make a wow experience for beginners. (They even partly know how does the *Meteor magic* work since it's clear that "to get the config ui you need some templates for it" by having this command instead of just an `accounts-facebook`).

The package name of the config UI may need further discussions too, while I suggest `accounts-facebook-config-ui` which is clear enough, I think.

**Comment by MeKarina** [comment]
@mitar ahhh i see i see never imagined such scenario

@wexpolyu
Yepp,,everything has pros and cons... 

We could make it back compat for the cost of making system a little too complex... We could make it simpler for the cost of breaking change and depreciating package or functionality..

**Comment by mitar** [comment]
@MeKarina: No worries. I also realized it only once I started replying to your question. Keep up with good work, proposals, and questions.

**Comment by laosb** [comment]
Deprecating `facebook` shouldn't have backcompat issues, since `facebook` will still do what it used to do but with a deprecation notice which leads users to use `facebook-oauth` instead.

**Comment by tmeasday** [comment]
Ok, so how does everyone feel about this solution:

1. Config UI removed from `{service}` -> `{service}-config-ui`
2. Remainder of `{service}` -> `{service}-oauth`, with old package implying it w/ deprecation warning
3. `accounts-{service}` depends only on `{service}-oauth`, but does log the warning above if the user is also using `accounts-ui` but not `{service}-config-ui`

**Comment by tmeasday** [comment]
(I think this probably a proposal made by @laosb before :) )

**Comment by laosb** [comment]
Almost, I guess.

**Comment by tmeasday** [comment]
Can people let us know if they have objections to this proposal? Otherwise I'm going to mark it PRs-encouraged.

**Comment by laosb** [comment]
Maybe we should have another label called `suggestions-encouraged` or something like that?

**Comment by tmeasday** [comment]
I think at some point someone has to just make a decision. I guess that's my job :/

**Comment by laosb** [comment]
So I should update my existing PR and reopen it, or just open a new one? I will work on it tonight(10hrs later), since I'll have a day off from school this afternoon.

**Comment by laosb** [comment]
I reopened #7728 and updated to your last proposal @tmeasday

**Comment by tmeasday** [comment]
The PR is now ready to merge if anyone would like to test it out a bit.

**Comment by tmeasday** [comment]
#7728 is merged!

**Comment by mitar** [comment]
Reopening because only Facebook has been split. Twitter (and Google) should probably be split as well.

**Comment by tmeasday** [comment]
Good point, thanks @mitar

**Comment by mitar** [comment]
But I would suggest that they are done in exactly the same manner. :-)

**Comment by queso** [comment]
I assume PRs are welcome for that, just following the other PR as a
template?

**Comment by tmeasday** [comment]
Yep! Although potentially we should wait until the FB changes have made it into a release to see if there are any big problems people run into.

**Comment by zeroasterisk** [comment]
Thanks very much for this work - I've been trying to track down the blaze/template dependancies for the client, and you guys are already on it.  THANKS!  

----

I, personally, am waiting on facebook, google, and twitter and those are the only things putting blaze into my (react UI) app.  c'mon and save me many KB!

**Comment by Method-X** [comment]
Any movement on this?

**Comment by laosb** [comment]
The facebook one should be released in 1.4.3 while others are not separated. Personally I'm having no time working on it.

**Comment by abernix** [comment]
@zeroasterisk @Method-X I see you're both actively interested in the other providers being ported over (in addition to Facebook).  Would either of you be interested in making your first code contributions to Meteor to make this happen?  It's mainly a matter of following the pattern of #7728 (Facebook) and therefore a great candidate for a first contribution that's not too complex.

**Comment by abernix** [comment]
With `accounts-twitter` and `accounts-google` completed in the pipeline now, it might be worth considering if this change should implement [an upgrader](https://github.com/meteor/meteor/blob/devel/tools/upgraders.js#L10-L20) which automatically makes changes to your `.meteor/packages`, adding new and removing deprecated packages as necessary.

Also, if anyone wanted to tackle the remaining `accounts-*` packages, `accounts-meteor-developer`, `accounts-github` and `accounts-weibo`, it'd be lovely to get these all into one release!  Otherwise, the finished ones are likely to land in 1.4.3.

**Comment by hwillson** [comment]
If no-one else has started it, I'll tackle `accounts-github`.

**Comment by laosb** [comment]
`accounts-weibo` is basically a China-based component. Let me do it if no one else has been working on it.

**Comment by abernix** [comment]
@laosb and @hwillson Fairly sure they are both up for grabs.  If feasible and you can get those done in the next couple days, we can land this entire issue in Meteor 1.4.3 – the only remaining task would be writing an upgrader that switches these packages automatically for the users.  This will ensure that `google` becomes `accounts-google-ui` AND `google-oauth` (and the same for the other packages).  It would be nice if we could do this in a single upgrader, rather than multiple.

**Comment by Method-X** [comment]
@abernix I've been learning to code since May. If I were more experienced I would help with this in a heartbeat.

**Comment by abernix** [comment]
@Method-X Realistically, the changes here are pretty straight forward and should (almost?) exactly replicate [these changes from the `google` change](https://github.com/meteor/meteor/pull/8283/files).  You'd need to analyze that PR's changes and be familiar with [running Meteor from checkout](https://github.com/meteor/meteor/#slow-start-for-developers).  That might be a task in and of itself for you to practice though!  That being said, @laosb and @hwillson have already volunteered here so I don't want to double up on efforts until they've opted out.  I'd ask that we consider the previously mentioned timeline (by end of Wednesday, perhaps) too though. :)

**Comment by abernix** [comment]
If anyone is able to test and isn't using `accounts-github` or `accounts-weibo`, I think this is ready to test in Meteor 1.4.3-beta.3:

    meteor update --release 1.4.3-beta.3

Input would be appreciated!

**Comment by hwillson** [comment]
I'll have `accounts-github` done first thing tomorrow morning - thanks!

**Comment by laosb** [comment]
Let me do it now!

**Comment by hwillson** [comment]
`accounts-github` changes are ready (PR #8303) and verified working with both Blaze and React. Thanks!

**Comment by abernix** [comment]
Thank you both, @laosb & @hwillson.  Also, I just realized `accounts-meteor-developer` is up for grabs still!  I'll do this myself if it comes down to the wire.

Also, can anyone verify my thinking on the upgrader (that is to say, the automated process of upgrading to this new system) here?  In looking at my previous comment, I think it's actually a bit more complex.  We want to ensure that there is zero broken experience for someone upgrading to Meteor 1.4.3 from previous Meteor versions.

I believe the following transformations should be made on existing apps, where `<provider>` is `google`, `twitter`, `github`, `weibo` or `meteor-developer`:
* Existence of `accounts-<provider>` **adds** `<provider>-config-ui`
* Existence of `<provider>` **changes to** `<provider>-oauth`

And I think the rest stays the same?

**Comment by hwillson** [comment]
I'll tackle `accounts-meteor-developer` - I should have it ready by tomorrow morning. Oh, what about `accounts-meetup`? I think we'll need to tackle that one as well. I can't commit to `accounts-meetup` before end of day tomorrow, but I'll post back if it looks like I'll have time.

Regarding your plan on the upgrader - yes, makes complete sense to me. You transformations look right - maybe just add `meetup` to the list (assuming it's done in time).

**Comment by laosb** [comment]
Do we still maintain `accounts-meetup`? If we don't, I guess deprecating it like deprecating `bootstrap` would be fine.

**Comment by hwillson** [comment]
`accounts-meteor-developer` should now be done (see #8305). Good question regarding `accounts-meetup` @laosb; as far as I know it's still maintained, but I'm not sure how much it's used. Deprecating it might be the way to go.

**Comment by hwillson** [comment]
@abernix - what do you think about `accounts-meetup`; should we refactor it or deprecate it? Either way I can work on this today, so let me know which direction you prefer. Thanks!

**Comment by abernix** [comment]
@hwillson With 931 lifetime installs, it's the lowest of the `accounts-*` packages and I'd be hard-pressed to say it's worth the investment of time to update it.  If it's super easy, I guess you can do it, but I fully support deprecating it.  Those want it still could still use the old version – granted it would load Blaze in, but probably a very, very small percentage of people doing that.  Deprecation PR gladly accepted in a similar manner to `code-prettify` (#8265).  If there is actually any backlash, we can un-deprecate it and ask those who need it to make the (relatively) simple changes.

**Comment by hwillson** [comment]
Makes sense to me @abernix - I'll make the necessary deprecation changes. Thanks!

**Comment by abernix** [comment]
Thank you, @hwillson!

**Comment by hwillson** [comment]
@abernix @laosb - I'm starting to re-think the idea of deprecating `accounts-meetup`. It isn't currently used by many developers, but the [Meteor Meetup community](https://www.meetup.com/topics/meteor/) is quite active. I've done a bit of digging around GitHub for users of `accounts-meetup`, and there are definitely up to date / current projects using it. Given that it's not that much work to update it, I think I'll move ahead with updating it instead of deprecating it. It's already being supported under the Meteor core umbrella (and hasn't really logged any issues in quite some time), so we're not really adding a lot of maintenance time by having it continue to be supported (once these OAuth changes are done that is). I'll submit a PR with these changes shortly (unless of course you guys object! 🙂). Thanks!

**Comment by hwillson** [comment]
PR #8321 contains the `accounts-meetup` changes. Let me know if there is anything else outstanding I can help with, to get this issue resolved. Thanks!

**Comment by stubailo** [comment]
@abernix is this closed now?

**Comment by abernix** [comment]
Yes!  This splitting of `accounts-*` packages has been released in Meteor 1.4.3.1.  While not officially `recommended` yet (In other words: developers receiving "update" notifications.  Coming soon!), you can test this now by upgrading to Meteor 1.4.3.1:

    meteor update --release 1.4.3.1

1.4.3.1 has an automatic upgrader in place which should split packages in `.meteor/packages` automatically.  Currently, it errs on the side of caution and tries to keep the app in the exact same state it was but if Blaze is not needed developers can remove the `{service}-config-ui` package from their apps with `meteor remove <package>`.

Just as a reminder, Blaze might still exist on the server-side (for boilerplate HTML reasons), however if you do not actually use Blaze as your view layer on the client, it will no longer be included in the client bundle.  This should be another milestone for reducing client bundle sizes for those using other view layers (React, Angular, etc.).

I'd like to give a huge shoutout to @laosb and @hwillson for submitting the (many!) PRs that made this all happen, and a big thanks to everyone else who participated in the design discussion as well.
